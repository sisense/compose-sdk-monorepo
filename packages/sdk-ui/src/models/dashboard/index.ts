export * from './types';
export * from './get-dashboard-model';
export * from './get-dashboard-models';
export * from './use-get-dashboard-model';
export * from './use-get-dashboard-models';
export * from './dashboard-model';
export * as dashboardModelTranslator from './dashboard-model-translator';
