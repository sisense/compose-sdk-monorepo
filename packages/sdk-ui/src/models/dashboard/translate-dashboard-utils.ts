import {
  type Layout as SisenseLayout,
  type FilterDto,
  type CascadingFilterDto,
  isCascadingFilterDto,
} from '../../api/types/dashboard-dto';
import type { Layout, WidgetFilterOptions } from './types';
import { CascadingFilter, Filter, createFilterFromJaql } from '@sisense/sdk-data';
import { WidgetDto } from '@/dashboard-widget/types';
import { CommonFiltersApplyMode } from '@/common-filters/types';

export const translateLayout = (layout: SisenseLayout): Layout => ({
  columns: layout.columns.map((c) => ({
    widthPercentage: c.width,
    rows: (c.cells || []).map((cell) => {
      const totalWidth = cell.subcells.reduce((acc, subcell) => acc + subcell.width, 0);
      return {
        cells: cell.subcells.map((subcell) => ({
          // If the total width of the subcells is less than 100, we increase width percentage to make the subcells fill the column
          widthPercentage: totalWidth < 100 ? subcell.width / (totalWidth / 100) : subcell.width,
          height: subcell.elements[0].height,
          widgetId: subcell.elements[0].widgetid,
        })),
      };
    }),
  })),
});

const createFilterFromFilterDto = (filterDto: FilterDto): Filter => {
  const filter: Filter = createFilterFromJaql(filterDto.jaql, filterDto.instanceid);
  filter.disabled = filterDto.disabled ?? false;
  return filter;
};

const createFilterFromCascadingFilterDto = (
  cascadingFilterDto: CascadingFilterDto,
): CascadingFilter => {
  const innerFilters = cascadingFilterDto.levels.map((level) => createFilterFromJaql(level));
  const filter = new CascadingFilter(innerFilters);
  filter.disabled = cascadingFilterDto.disabled ?? false;
  return filter;
};

export function extractDashboardFilters(
  dashboardFilters: Array<FilterDto | CascadingFilterDto>,
): Filter[] {
  return dashboardFilters.map((f) =>
    isCascadingFilterDto(f) ? createFilterFromCascadingFilterDto(f) : createFilterFromFilterDto(f),
  );
}

export function translateWidgetFilterOptions(widgets: WidgetDto[] = []): WidgetFilterOptions {
  const widgetFilterOptionsMap: WidgetFilterOptions = {};

  widgets.forEach((widget: WidgetDto) => {
    widgetFilterOptionsMap[widget.oid] = {
      applyMode:
        widget.options?.dashboardFiltersMode === 'filter'
          ? CommonFiltersApplyMode.FILTER
          : CommonFiltersApplyMode.HIGHLIGHT,
      shouldAffectFilters: widget.options?.selector,
      ignoreFilters: {
        all: widget.metadata.ignore?.all,
        ids: widget.metadata.ignore?.ids,
      },
    };
  });

  return widgetFilterOptionsMap;
}
