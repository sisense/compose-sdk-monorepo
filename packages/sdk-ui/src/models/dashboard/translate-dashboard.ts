import { type DashboardDto } from '../../api/types/dashboard-dto';
import { translateWidget } from '../widget/translate-widget';
import { type DashboardModel } from './types';

export function translateDashboard({
  oid,
  title,
  datasource,
  widgets,
}: DashboardDto): DashboardModel {
  return {
    oid,
    title,
    dataSource: datasource.fullname ?? datasource.title,
    ...(widgets && {
      widgets: widgets.map(translateWidget),
    }),
  };
}
