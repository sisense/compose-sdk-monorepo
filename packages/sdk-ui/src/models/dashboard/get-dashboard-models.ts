import { HttpClient } from '@sisense/sdk-rest-client';
import { RestApi } from '../../api/rest-api.js';
import { translateDashboard } from './translate-dashboard.js';

export interface GetDashboardModelsOptions {
  /**
   * Dashboard title to search by
   *
   * Dashboard titles are not necessarily unique, so the result may contain multiple dashboards.
   */
  searchByTitle?: string;
  /**
   * {@inheritDoc GetDashboardModelOptions.includeWidgets}
   */
  includeWidgets?: boolean;
}

/** @internal */
export async function getDashboardModels(
  http: HttpClient,
  options: GetDashboardModelsOptions = {},
) {
  const { includeWidgets, searchByTitle } = options;
  const api = new RestApi(http);
  const expand: string[] = [];
  let fields = ['oid', 'title', 'datasource'];

  if (includeWidgets) {
    // Removes "fields" list due to it's conflict with "expand" parameter => API endpoint issue
    fields = [];
    expand.push('widgets');
  }

  const dashboards = await api.getDashboards({ fields, expand, searchByTitle });

  return dashboards.map((dashboard) => translateDashboard(dashboard));
}
