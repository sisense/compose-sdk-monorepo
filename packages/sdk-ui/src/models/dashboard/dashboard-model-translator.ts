import { DashboardProps } from '@/dashboard';
import { DashboardModel } from './dashboard-model.js';
import { widgetModelTranslator } from '@/models/widget/';
import { DashboardDto } from '@/api/types/dashboard-dto.js';
import { CompleteThemeSettings } from '@/types.js';
import { AppSettings } from '@/app/settings/settings.js';
import {
  extractDashboardFilters,
  translateLayout,
  translateTabbersOptions,
  translateWidgetsOptions,
} from './translate-dashboard-utils.js';
import { DataSource } from '@sisense/sdk-data';

/**
 * Translates {@link DashboardModel} to {@link DashboardProps}.
 *
 * @example
 * ```tsx
 * <Dashboard {...dashboardModelTranslator.toDashboardProps(dashboardModel)} />
 * ```
 * @group Fusion Assets
 * @fusionEmbed
 */
export function toDashboardProps(dashboardModel: DashboardModel): DashboardProps {
  const {
    title,
    dataSource,
    widgets: widgetModels,
    layoutOptions,
    filters,
    widgetsOptions,
    tabbersOptions,
    styleOptions,
  } = dashboardModel;
  return {
    title,
    defaultDataSource: dataSource,
    widgets: widgetModels.map(widgetModelTranslator.toWidgetProps),
    layoutOptions,
    config: {},
    filters,
    widgetsOptions,
    styleOptions,
    tabbersOptions,
  };
}

/**
 * Creates a new dashboard model from a dashboard DTO.
 *
 * @param dashboardDto - The dashboard DTO to be converted to a dashboard model
 * @param themeSettings - Optional theme settings
 * @param appSettings - Optional application settings
 * @internal
 */
export function fromDashboardDto(
  dashboardDto: DashboardDto,
  themeSettings?: CompleteThemeSettings,
  appSettings?: AppSettings,
): DashboardModel {
  const {
    oid,
    title,
    datasource,
    widgets: widgetDtoList,
    layout: layoutDto,
    filters: filterDtoList,
    filterRelations: filterRelationsDtoOptions,
    style,
  } = dashboardDto;

  const dataSource: DataSource = {
    title: datasource.title,
    type: datasource.live ? 'live' : 'elasticube',
  };
  const styleOptions = {
    ...(style?.palette ? { palette: { variantColors: style?.palette.colors } } : null),
  };

  const mergedThemeSettings = themeSettings
    ? {
        ...themeSettings,
        ...(styleOptions.palette ? { palette: styleOptions.palette } : null),
      }
    : themeSettings;
  const widgets =
    widgetDtoList?.map((widget) =>
      widgetModelTranslator.fromWidgetDto(widget, mergedThemeSettings, appSettings),
    ) || [];
  const widgetsPanelLayout = layoutDto ? translateLayout(layoutDto) : { columns: [] };
  const filterRelationsModel = filterRelationsDtoOptions?.[0]?.filterRelations;
  const filters = extractDashboardFilters(filterDtoList || [], filterRelationsModel);
  const widgetsOptions = translateWidgetsOptions(widgetDtoList);
  const tabbersOptions = translateTabbersOptions(widgetDtoList);

  const dashboardModel: DashboardModel = {
    oid,
    title,
    dataSource,
    styleOptions,
    widgets,
    layoutOptions: { widgetsPanel: widgetsPanelLayout },
    filters,
    widgetsOptions,
    tabbersOptions,
  };

  return dashboardModel;
}
