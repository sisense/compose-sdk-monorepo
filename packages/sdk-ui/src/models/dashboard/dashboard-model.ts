import { DataSource, Filter, FilterRelations } from '@sisense/sdk-data';
import { DashboardStyleOptions, TabbersOptions, WidgetModel } from '@/models';
import { type WidgetsOptions } from './types';
import { DashboardLayoutOptions } from '@/dashboard';

/**
 * Model of Sisense Fusion dashboard defined in the abstractions of Compose SDK.
 *
 * @group Fusion Assets
 * @fusionEmbed
 */
export interface DashboardModel {
  /**
   * Unique identifier of the dashboard.
   */
  oid: string;

  /**
   * Dashboard title.
   */
  title: string;

  /**
   * Dashboard data source details.
   */
  dataSource: DataSource;

  /**
   * Dashboard widget models.
   */
  widgets: WidgetModel[];

  /**
   * Dashboard layout options.
   *
   * @internal
   */
  layoutOptions: DashboardLayoutOptions;

  /**
   * Dashboard style options.
   *
   * @internal
   */
  styleOptions: DashboardStyleOptions;

  /**
   * Dashboard filters.
   *
   * @internal
   */
  filters: Filter[] | FilterRelations;

  /**
   * Dashboard options for each of the widgets.
   *
   * @internal
   */
  widgetsOptions: WidgetsOptions;

  /**
   * Tabber configurations for the dashboard.
   *
   * @internal
   */
  tabbersOptions?: TabbersOptions;
}
