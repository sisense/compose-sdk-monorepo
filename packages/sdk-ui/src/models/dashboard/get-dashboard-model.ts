import { HttpClient } from '@sisense/sdk-rest-client';
import { RestApi } from '../../api/rest-api';
import { DashboardModel, dashboardModelTranslator } from '@/models/dashboard';
import { CompleteThemeSettings } from '../../types';
import { AppSettings } from '@/app/settings/settings';
import { TranslatableError } from '@/translation/translatable-error';
import { getWidgetIdsFromDashboard } from '@/utils/extract-widget-ids';
import { WidgetDto } from '@/widget-by-id/types';

export interface GetDashboardModelOptions {
  /**
   * Boolean flag whether to include widgets in the dashboard model
   *
   * If not specified, the default value is `false`
   */
  includeWidgets?: boolean;

  /**
   * Boolean flag whether to include filters in the dashboard model
   *
   * If not specified, the default value is `false`
   */
  includeFilters?: boolean;
}

/**
 * Retrieves a dashboard model by its OID.
 *
 * @param http - The HTTP client
 * @param dashboardOid - The OID of the dashboard
 * @param options - The options to include widgets and filters in the dashboard model
 * @param themeSettings - Optional theme settings
 * @param appSettings - Optional application settings
 * @returns The dashboard model
 * @internal
 */
export async function getDashboardModel(
  http: HttpClient,
  dashboardOid: string,
  options: GetDashboardModelOptions = {},
  themeSettings?: CompleteThemeSettings,
  appSettings?: AppSettings,
): Promise<DashboardModel> {
  const { includeWidgets, includeFilters } = options;
  const api = new RestApi(http);
  const fields = ['oid', 'title', 'datasource', 'style'];

  const isWat = http.auth?.type === 'wat';

  if (includeWidgets) {
    fields.push('layout');
  }

  if (includeFilters) {
    fields.push('filters');
    fields.push('filterRelations');
  }

  const dashboard = await api.getDashboard(dashboardOid, { fields });

  if (!dashboard) {
    throw new TranslatableError('errors.dashboardWithOidNotFound', { dashboardOid });
  }

  if (includeWidgets) {
    let widgets: WidgetDto[];

    if (isWat) {
      // WAT (Web Access Token) authentication method restricts direct access to the
      // API endpoint `api/v1/dashboards/${dashboardOid}/widgets`.
      // As a workaround, widgets are fetched individually based on their references in the dashboard layout.
      const widgetIds = getWidgetIdsFromDashboard(dashboard);
      const fetchedWidgets = await Promise.all(
        widgetIds.map((id) => api.getWidget(id, dashboard.oid)),
      );
      widgets = fetchedWidgets.filter((widget): widget is WidgetDto => widget !== undefined);
    } else {
      // Fetch all widgets at once
      widgets = (await api.getDashboardWidgets(dashboardOid)) || [];
    }

    dashboard.widgets = widgets;
  }

  // Next could be replaced in future with expand 'style' in '/dashboards/' request
  // when lowest supported Sisense API version will support it.
  // Currently, expand 'style' work only from l2024.2.0 and cause crash in older.
  if (dashboard.style?.paletteId && !dashboard.style.palette) {
    const palettesDto = await api.getPalettes();
    const paletteDto = palettesDto?.find(({ _id }) => _id === dashboard.style?.paletteId);
    if (paletteDto) {
      dashboard.style.palette = {
        name: paletteDto.name,
        colors: paletteDto.colors,
      };
    }
  }

  return dashboardModelTranslator.fromDashboardDto(dashboard, themeSettings, appSettings);
}
