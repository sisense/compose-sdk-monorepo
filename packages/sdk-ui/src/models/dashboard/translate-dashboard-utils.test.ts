import { RestApi } from '@/api/rest-api';
import { translateLayout, withSharedFormulas } from '@/models/dashboard/translate-dashboard-utils';
import isEqual from 'lodash-es/isEqual';
import {
  dashboardWithSharedFormulas,
  sharedFormulasDictionary,
} from '../__mocks__/dashboard-with-shared-formulas';

describe('translate-dashboard-utils', () => {
  describe('translateLayout', () => {
    it('should correctly translate layout', () => {
      const dashboardDtoLayout = {
        instanceid: '78526-6FA5-87',
        type: 'columnar',
        columns: [
          {
            width: 50,
            cells: [
              {
                subcells: [
                  {
                    elements: [
                      {
                        minHeight: 64,
                        maxHeight: 1028,
                        height: '212px',
                        minWidth: 48,
                        maxWidth: 1028,
                        defaultWidth: 512,
                        widgetid: '6662ffcf888f5e002aa6cf18',
                      },
                    ],
                    width: 50,
                    stretchable: false,
                    pxlWidth: 515.5,
                    index: 0,
                  },
                  {
                    elements: [
                      {
                        minHeight: 64,
                        maxHeight: 1028,
                        height: '212px',
                        minWidth: 48,
                        maxWidth: 1028,
                        defaultWidth: 512,
                        widgetid: '66630139888f5e002aa6cf1a',
                      },
                    ],
                    width: 50,
                    stretchable: false,
                    pxlWidth: 515.5,
                    index: 1,
                  },
                ],
              },
              {
                subcells: [
                  {
                    elements: [
                      {
                        minHeight: 64,
                        maxHeight: 1028,
                        height: 192,
                        minWidth: 48,
                        maxWidth: 1028,
                        defaultWidth: 512,
                        widgetid: '66630141888f5e002aa6cf1c',
                      },
                    ],
                    width: 100,
                    stretchable: false,
                    pxlWidth: 1031,
                    index: 0,
                  },
                ],
              },
            ],
            pxlWidth: 1031,
            index: 0,
          },
          {
            width: 50,
          },
        ],
        container: {},
      };

      const result = translateLayout(dashboardDtoLayout);

      expect(result).toEqual({
        columns: [
          {
            widthPercentage: 50,
            rows: [
              {
                cells: [
                  { widthPercentage: 50, height: '212px', widgetId: '6662ffcf888f5e002aa6cf18' },
                  { widthPercentage: 50, height: '212px', widgetId: '66630139888f5e002aa6cf1a' },
                ],
              },
              {
                cells: [
                  { widthPercentage: 100, height: 192, widgetId: '66630141888f5e002aa6cf1c' },
                ],
              },
            ],
          },
          { widthPercentage: 50, rows: [] },
        ],
      });
    });

    it('should correctly translate layout with subcells that doesnt fit to column width in total', () => {
      const dashboardDtoLayout = {
        instanceid: '78526-6FA5-87',
        type: 'columnar',
        columns: [
          {
            width: 50,
            cells: [
              {
                subcells: [
                  {
                    elements: [
                      {
                        minHeight: 64,
                        maxHeight: 1028,
                        height: '212px',
                        minWidth: 48,
                        maxWidth: 1028,
                        defaultWidth: 512,
                        widgetid: '6662ffcf888f5e002aa6cf18',
                      },
                    ],
                    width: 25,
                    stretchable: false,
                    pxlWidth: 515.5,
                    index: 0,
                  },
                  {
                    elements: [
                      {
                        minHeight: 64,
                        maxHeight: 1028,
                        height: '212px',
                        minWidth: 48,
                        maxWidth: 1028,
                        defaultWidth: 512,
                        widgetid: '66630139888f5e002aa6cf1a',
                      },
                    ],
                    width: 25,
                    stretchable: false,
                    pxlWidth: 515.5,
                    index: 1,
                  },
                ],
              },
              {
                subcells: [
                  {
                    elements: [
                      {
                        minHeight: 64,
                        maxHeight: 1028,
                        height: 192,
                        minWidth: 48,
                        maxWidth: 1028,
                        defaultWidth: 512,
                        widgetid: '66630141888f5e002aa6cf1c',
                      },
                    ],
                    width: 10,
                    stretchable: false,
                    pxlWidth: 1031,
                    index: 0,
                  },
                ],
              },
            ],
            pxlWidth: 1031,
            index: 0,
          },
          {
            width: 50,
          },
        ],
        container: {},
      };

      const result = translateLayout(dashboardDtoLayout);

      expect(result).toEqual({
        columns: [
          {
            widthPercentage: 50,
            rows: [
              {
                cells: [
                  { widthPercentage: 50, height: '212px', widgetId: '6662ffcf888f5e002aa6cf18' },
                  { widthPercentage: 50, height: '212px', widgetId: '66630139888f5e002aa6cf1a' },
                ],
              },
              {
                cells: [
                  { widthPercentage: 100, height: 192, widgetId: '66630141888f5e002aa6cf1c' },
                ],
              },
            ],
          },
          { widthPercentage: 50, rows: [] },
        ],
      });
    });

    it('should correctly translate empty layout', () => {
      const dashboardDtoLayout = {};

      const result = translateLayout(dashboardDtoLayout);

      expect(result).toEqual({
        columns: [],
      });
    });
  });

  describe('withSharedFormulas', () => {
    const api = {
      getSharedFormulas: vi.fn(),
    };

    beforeEach(() => {
      vi.clearAllMocks();
    });

    it('should correctly replace shared formulas', async () => {
      api.getSharedFormulas.mockImplementation((sharedFormulasIds: string[]) => {
        if (isEqual(sharedFormulasIds, Object.keys(sharedFormulasDictionary))) {
          return sharedFormulasDictionary;
        } else {
          throw new Error('Invalid shared formulas ids');
        }
      });

      const result = await withSharedFormulas(
        dashboardWithSharedFormulas,
        api as unknown as RestApi,
      );
      expect(result).toMatchSnapshot();
      expect(api.getSharedFormulas).toHaveBeenCalledWith(Object.keys(sharedFormulasDictionary));
    });
  });
});
