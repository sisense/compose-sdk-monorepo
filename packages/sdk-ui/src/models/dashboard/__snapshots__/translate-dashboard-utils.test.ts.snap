// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translate-dashboard-utils > withSharedFormulas > should correctly replace shared formulas 1`] = `
{
  "datasource": {
    "address": "LocalHost",
    "database": "aSampleIAAaECommerce",
    "fullname": "localhost/Sample ECommerce",
    "id": "localhost_aSampleIAAaECommerce",
    "live": false,
    "title": "Sample ECommerce",
  },
  "filters": [],
  "layout": {
    "columns": [
      {
        "cells": [
          {
            "subcells": [
              {
                "elements": [
                  {
                    "defaultWidth": 512,
                    "height": 384,
                    "maxHeight": 2048,
                    "maxWidth": 2048,
                    "minHeight": 96,
                    "minWidth": 128,
                    "widgetid": "675325cb65ad730034b436f1",
                  },
                ],
                "index": 0,
                "pxlWidth": 1543.21,
                "stretchable": false,
                "width": 100,
              },
            ],
          },
          {
            "subcells": [
              {
                "elements": [
                  {
                    "defaultWidth": 512,
                    "height": 192,
                    "maxHeight": 1028,
                    "maxWidth": 1028,
                    "minHeight": 64,
                    "minWidth": 48,
                    "widgetid": "6758464e65ad730034b43751",
                  },
                ],
              },
            ],
          },
        ],
        "index": 0,
        "pxlWidth": 1543.21,
        "width": 100,
      },
    ],
    "container": {},
    "instanceid": "39478-DA23-28",
    "type": "columnar",
  },
  "oid": "675325cb65ad730034b436f0",
  "style": {},
  "title": "shared-formula (1)",
  "widgets": [
    {
      "_id": "6753266ae2d924c029c93dd2",
      "created": "2024-12-06T16:26:51.444Z",
      "dashboardid": "675325cb65ad730034b436f0",
      "datasource": {
        "address": "LocalHost",
        "database": "aSampleIAAaECommerce",
        "fullname": "localhost/Sample ECommerce",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "desc": null,
      "instanceType": "owner",
      "instanceid": "C784E-1DC1-3A",
      "lastOpened": null,
      "lastUpdated": "2024-12-10T13:46:30.590Z",
      "metadata": {
        "ignore": {
          "all": false,
          "dimensions": [],
          "ids": [],
        },
        "panels": [
          {
            "items": [
              {
                "field": {
                  "id": "[Commerce.Age Range]",
                  "index": 0,
                },
                "format": {},
                "instanceid": "995C0-E51E-8C",
                "jaql": {
                  "column": "Age Range",
                  "datatype": "text",
                  "dim": "[Commerce.Age Range]",
                  "merged": true,
                  "table": "Commerce",
                  "title": "Age Range",
                },
                "panel": "rows",
              },
            ],
            "name": "categories",
          },
          {
            "items": [
              {
                "format": {
                  "color": {
                    "colorIndex": 0,
                    "type": "color",
                  },
                  "mask": {
                    "abbreviateAll": false,
                    "abbreviations": {
                      "b": true,
                      "k": true,
                      "m": true,
                      "t": true,
                    },
                    "decimals": "auto",
                    "isdefault": true,
                    "separated": true,
                    "type": "number",
                  },
                },
                "instanceid": "EA32A-6C7E-D5",
                "jaql": {
                  "context": {
                    "[4525D-D11]": {
                      "context": {
                        "[388A1-BCC]": {
                          "column": "Cost",
                          "datatype": "numeric",
                          "dim": "[Commerce.Cost]",
                          "table": "Commerce",
                          "title": "Cost",
                        },
                        "[4E7D2-8BA]": {
                          "column": "Quantity",
                          "datatype": "numeric",
                          "dim": "[Commerce.Quantity]",
                          "table": "Commerce",
                          "title": "Quantity",
                        },
                      },
                      "created": "2024-12-06T16:30:00.204Z",
                      "datasourceTitle": "Sample ECommerce",
                      "description": "",
                      "formula": "AVG(CORREL([388A1-BCC],[4E7D2-8BA]))",
                      "lastUpdated": "2024-12-06T16:30:00.204Z",
                      "lastUpdatedUser": "662ba7319f04e5001cbc7f58",
                      "oid": "e78303f1-820c-4a02-91ab-6d2b8700abeb",
                      "owner": "662ba7319f04e5001cbc7f58",
                      "title": "average-correlation",
                    },
                  },
                  "formula": "[4525D-D11]",
                  "title": "[average-correlation]",
                  "type": "measure",
                },
                "panel": "measures",
              },
            ],
            "name": "values",
          },
          {
            "items": [],
            "name": "break by",
          },
          {
            "items": [],
            "name": "filters",
          },
        ],
        "usedFormulasMapping": {
          "e78303f1-820c-4a02-91ab-6d2b8700abeb": [
            "metadata.panels[1].items[0].jaql.context['[4525D-D11]']",
          ],
        },
      },
      "oid": "675325cb65ad730034b436f1",
      "options": {
        "autoUpdateOnEveryChange": true,
        "dashboardFiltersMode": "select",
        "drillToAnywhere": true,
        "previousScrollerLocation": {
          "max": null,
          "min": null,
        },
        "selector": true,
        "triggersDomready": true,
      },
      "owner": "662ba7319f04e5001cbc7f58",
      "realTimeRefreshing": false,
      "selection": null,
      "source": null,
      "style": {
        "dataLimits": {
          "categoriesCapacity": 100000,
          "seriesCapacity": 50,
        },
        "legend": {
          "enabled": true,
          "position": "bottom",
        },
        "narration": {
          "aggregation": "sum",
          "display": "above",
          "enabled": false,
          "format": "bullets",
          "labels": [
            {
              "id": "age_range",
              "plural": "Age Range",
              "singular": "Age Range",
              "title": "Age Range",
            },
          ],
          "up_sentiment": "good",
          "verbosity": "medium",
        },
        "navigator": {
          "enabled": true,
        },
        "seriesLabels": {
          "enabled": false,
          "labels": {
            "enabled": false,
            "stacked": false,
            "stackedPercentage": false,
            "types": {
              "count": false,
              "percentage": false,
              "relative": false,
              "totals": false,
            },
          },
          "rotation": 0,
        },
        "xAxis": {
          "enabled": true,
          "gridLines": true,
          "isIntervalEnabled": false,
          "labels": {
            "enabled": true,
            "rotation": 0,
          },
          "ticks": true,
          "title": {
            "enabled": false,
          },
          "x2Title": {
            "enabled": false,
          },
        },
        "y2Axis": {
          "enabled": true,
          "gridLines": false,
          "hideMinMax": false,
          "inactive": true,
          "isIntervalEnabled": true,
          "labels": {
            "enabled": true,
            "rotation": 0,
          },
          "logarithmic": false,
          "ticks": true,
          "title": {
            "enabled": false,
          },
        },
        "yAxis": {
          "enabled": true,
          "gridLines": true,
          "hideMinMax": false,
          "inactive": false,
          "isIntervalEnabled": true,
          "labels": {
            "enabled": true,
            "rotation": 0,
          },
          "logarithmic": false,
          "ticks": true,
          "title": {
            "enabled": false,
          },
        },
      },
      "subtype": "column/classic",
      "title": "1",
      "type": "chart/column",
      "userId": "662ba7319f04e5001cbc7f58",
    },
    {
      "_id": "675850ade2d924c0294b1acf",
      "created": "2024-12-10T13:46:54.908Z",
      "dashboardid": "675325cb65ad730034b436f0",
      "datasource": {
        "address": "LocalHost",
        "database": "aSampleIAAaECommerce",
        "fullname": "localhost/Sample ECommerce",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "desc": null,
      "instanceType": "owner",
      "instanceid": "95FBA-2682-D8",
      "lastUpdated": "2024-12-10T14:31:09.923Z",
      "metadata": {
        "ignore": {
          "all": false,
          "dimensions": [],
          "ids": [],
        },
        "panels": [
          {
            "items": [
              {
                "format": {
                  "color": {
                    "colorIndex": 0,
                    "type": "color",
                  },
                  "mask": {
                    "abbreviateAll": false,
                    "abbreviations": {
                      "b": true,
                      "k": false,
                      "m": true,
                      "t": true,
                    },
                    "decimals": "auto",
                    "isdefault": true,
                    "separated": true,
                    "type": "number",
                  },
                },
                "instanceid": "CBFDE-BDC5-AF",
                "jaql": {
                  "context": {
                    "[5A349-31A]": {
                      "context": {
                        "[5CD73-E54]": {
                          "agg": "sum",
                          "column": "Cost",
                          "datatype": "numeric",
                          "dim": "[Commerce.Cost]",
                          "table": "Commerce",
                          "title": "Total Cost",
                        },
                      },
                      "created": "2024-12-06T16:03:16.223Z",
                      "datasourceTitle": "Sample ECommerce",
                      "description": "",
                      "formula": "LN([5CD73-E54])",
                      "lastUpdated": "2024-12-06T16:03:16.223Z",
                      "lastUpdatedUser": "662ba7319f04e5001cbc7f58",
                      "oid": "6d3a7c78-e2f6-4233-bac2-adda493e0e37",
                      "owner": "662ba7319f04e5001cbc7f58",
                      "title": "logarifm of cost",
                    },
                  },
                  "formula": "[5A349-31A]",
                  "title": "[logarifm of cost]",
                  "type": "measure",
                },
              },
            ],
            "name": "value",
          },
          {
            "items": [
              {
                "format": {
                  "mask": {
                    "abbreviateAll": false,
                    "abbreviations": {
                      "b": true,
                      "k": false,
                      "m": true,
                      "t": true,
                    },
                    "decimals": "auto",
                    "isdefault": true,
                    "separated": true,
                    "type": "number",
                  },
                },
                "instanceid": "72297-CF53-33",
                "jaql": {
                  "context": {
                    "[02006-0AB]": {
                      "context": {
                        "[388A1-BCC]": {
                          "column": "Cost",
                          "datatype": "numeric",
                          "dim": "[Commerce.Cost]",
                          "table": "Commerce",
                          "title": "Cost",
                        },
                        "[4E7D2-8BA]": {
                          "column": "Quantity",
                          "datatype": "numeric",
                          "dim": "[Commerce.Quantity]",
                          "table": "Commerce",
                          "title": "Quantity",
                        },
                      },
                      "created": "2024-12-06T16:30:00.204Z",
                      "datasourceTitle": "Sample ECommerce",
                      "description": "",
                      "formula": "AVG(CORREL([388A1-BCC],[4E7D2-8BA]))",
                      "lastUpdated": "2024-12-06T16:30:00.204Z",
                      "lastUpdatedUser": "662ba7319f04e5001cbc7f58",
                      "oid": "e78303f1-820c-4a02-91ab-6d2b8700abeb",
                      "owner": "662ba7319f04e5001cbc7f58",
                      "title": "average-correlation",
                    },
                  },
                  "formula": "[02006-0AB]",
                  "loadPath": true,
                  "title": "[average-correlation]",
                  "type": "measure",
                },
              },
            ],
            "name": "secondary",
          },
          {
            "items": [],
            "name": "min",
          },
          {
            "items": [],
            "name": "max",
          },
          {
            "items": [],
            "name": "filters",
          },
        ],
        "usedFormulasMapping": {
          "6d3a7c78-e2f6-4233-bac2-adda493e0e37": [
            "metadata.panels[0].items[0].jaql.context['[5A349-31A]']",
          ],
          "e78303f1-820c-4a02-91ab-6d2b8700abeb": [
            "metadata.panels[1].items[0].jaql.context['[02006-0AB]']",
          ],
        },
      },
      "oid": "6758464e65ad730034b43751",
      "options": {
        "autoUpdateOnEveryChange": true,
        "dashboardFiltersMode": "filter",
        "disallowSelector": true,
        "selector": false,
        "supportsHierarchies": false,
        "triggersDomready": true,
      },
      "owner": "662ba7319f04e5001cbc7f58",
      "realTimeRefreshing": false,
      "selection": null,
      "source": null,
      "style": {
        "components": {
          "icon": {
            "enabled": true,
            "inactive": true,
          },
          "secondaryTitle": {
            "enabled": true,
            "inactive": false,
          },
          "title": {
            "enabled": true,
            "inactive": false,
          },
        },
        "indicator/gauge": {
          "components": {
            "labels": {
              "enabled": true,
              "inactive": false,
            },
            "secondaryTitle": {
              "enabled": true,
              "inactive": true,
            },
            "ticks": {
              "enabled": true,
              "inactive": false,
            },
            "title": {
              "enabled": true,
              "inactive": false,
            },
          },
          "skin": "1",
          "subtype": "round",
        },
        "indicator/numeric": {
          "components": {
            "icon": {
              "enabled": true,
              "inactive": false,
            },
            "secondaryTitle": {
              "enabled": true,
              "inactive": true,
            },
            "title": {
              "enabled": true,
              "inactive": false,
            },
          },
          "skin": "vertical",
          "subtype": "simple",
        },
        "narration": {
          "aggregation": "sum",
          "display": "above",
          "enabled": false,
          "format": "bullets",
          "labels": [],
          "up_sentiment": "good",
          "verbosity": "medium",
        },
        "skin": "vertical",
        "subtype": "simple",
      },
      "subtype": "indicator/numeric",
      "tags": [],
      "title": "",
      "type": "indicator",
      "userId": "662ba7319f04e5001cbc7f58",
    },
  ],
}
`;
